### ISSUE TITLE
Validation loss increases after a certain epoch.

### ISSUE THREAD

**Issue by EugeneKok97** [hrlblab/CircleNet#20]
Hi, I realized that my validation loss is increasing over time after a certain epoch while the training loss is converging. It seems to me that the network is overfitting. I have tried disabling random cropping, and introduced augmentations to mitigate this issue, however the problem seems to remain. I'm fairly new to neural network training, any suggestion is much appreciated!

![Screenshot from 2022-11-17 11-39-49](https://user-images.githubusercontent.com/72436847/202326164-c22f0a57-79fb-42c7-9b97-2a142f1b6bc5.png)
![Screenshot from 2022-11-17 11-40-02](https://user-images.githubusercontent.com/72436847/202326167-49d1cb98-2b5b-434e-b761-7dd552df384d.png)

Here is my opt file:
```=> torch version: 0.4.1
==> cudnn version: 7104
==> Cmd:
['main.py', 'circledet', '--exp_id', 'Apple5', '--arch', 'hourglass', '--batch_size', '4', '--master_batch', '4', '--lr', '2.5e-4', '--dataset', 'apple2', '--load_model', '/home/lmga-titanx/CircleNet/src/lib/models/ctdet_coco_hg.pth', '--num_epochs', '1000', '--input_w', '128', '--input_h', '128', '--K', '1', '--not_rand_crop', '--rotate', '1']
==> Opt:
  K: 1
  aggr_weight: 0.0
  agnostic_ex: False
  arch: hourglass
  aug_ddd: 0.5
  aug_rot: 0
  batch_size: 4
  cat_spec_wh: False
  center_thresh: 0.1
  chunk_sizes: [4]
  data_dir: /home/lmga-titanx/CircleNet/src/lib/../../data
  dataset: apple2
  debug: 0
  debug_dir: /home/lmga-titanx/CircleNet/src/lib/../../exp/circledet/Apple5/debug
  debugger_theme: white
  demo: 
  demo_dir: 
  dense_hp: False
  dense_wh: False
  dep_weight: 1
  dim_weight: 1
  down_ratio: 4
  eval_oracle_dep: False
  eval_oracle_hm: False
  eval_oracle_hmhp: False
  eval_oracle_hp_offset: False
  eval_oracle_kps: False
  eval_oracle_offset: False
  eval_oracle_wh: False
  exp_dir: /home/lmga-titanx/CircleNet/src/lib/../../exp/circledet
  exp_id: Apple5
  ez_guassian_radius: False
  filter_boarder: False
  fix_res: True
  flip: 0.5
  flip_test: False
  gpus: [0]
  gpus_str: 0
  head_conv: 64
  heads: {'hm': 1, 'cl': 1, 'reg': 2}
  hide_data_time: False
  hm_hp: True
  hm_hp_weight: 1
  hm_weight: 1
  hp_weight: 1
  input_h: 128
  input_res: 128
  input_w: 128
  keep_res: False
  kitti_split: 3dop
  load_model: /home/lmga-titanx/CircleNet/src/lib/models/ctdet_coco_hg.pth
  lr: 0.00025
  lr_step: [90, 120]
  lv: 2
  mask_focal_loss: False
  master_batch_size: 4
  mean: [[[0.40789655 0.44719303 0.47026116]]]
  metric: loss
  mse_loss: False
  nms: False
  no_color_aug: False
  norm_wh: False
  not_cuda_benchmark: False
  not_hm_hp: False
  not_prefetch_test: False
  not_rand_crop: True
  not_reg_bbox: False
  not_reg_hp_offset: False
  not_reg_offset: False
  num_classes: 1
  num_epochs: 1000
  num_iters: -1
  num_stacks: 2
  num_workers: 4
  off_weight: 1
  ontestdata: False
  output_h: 32
  output_res: 32
  output_w: 32
  pad: 127
  peak_thresh: 0.2
  print_iter: 0
  rect_mask: False
  reg_bbox: True
  reg_hp_offset: True
  reg_loss: l1
  reg_offset: True
  resume: False
  root_dir: /home/lmga-titanx/CircleNet/src/lib/../..
  rot_weight: 1
  rotate: 1.0
  rotate_reproduce: 0
  save_all: False
  save_dir: /home/lmga-titanx/CircleNet/src/lib/../../exp/circledet/Apple5
  scale: 0.4
  scores_thresh: 0.1
  seed: 317
  shift: 0.1
  std: [[[0.2886383  0.27408165 0.27809834]]]
  task: circledet
  test: False
  test_scales: [1.0]
  test_val_name: 
  trainval: False
  val_intervals: 1
  vis_thresh: 0.2
  wh_weight: 0.1```

**Comment by EthanHNguyen** [comment]
No problem. This is a common issue with convolutional neural networks. This [article](https://www.analyticsvidhya.com/blog/2020/09/overfitting-in-cnn-show-to-treat-overfitting-in-convolutional-neural-networks/) covers some things you can do.

I would encourage you to use random cropping in addition to whatever augmentation you have. You can also add regularization to the loss function.

**Comment by EthanHNguyen** [comment]
Also, I would make sure that you're initializing your weights from pre-trained models. See the bottom of [this page](https://github.com/hrlblab/CircleNet/blob/master/docs/GETTING_STARTED.md)

**Comment by EugeneKok97** [comment]
@EthanHNguyen Thank you so much for your feedback! In regards to adding regularization, is there an option in ```opts``` that allows me to do so, or do I have to add new layers into the existing code?

I believe that I should've initialized the weights from pre-trained model by having ```load_model path/to/ctdet_coco_hg.pth``` , where ctdet_coco_hg.pth was downloaded from [this page](https://github.com/hrlblab/CircleNet/blob/master/docs/GETTING_STARTED.md). Please let me know if I'm missing something. Thank you!

**Comment by EthanHNguyen** [comment]
> @EthanHNguyen Thank you so much for your feedback! In regards to adding regularization, is there an option in `opts` that allows me to do so, or do I have to add new layers into the existing code?

No, there isn't a built-in option. In our experiments, we mainly used image augmentation and early stopping as regularization. We simply chose the model that performed best on validation.

I would add regularization to the loss function [here](https://github.com/hrlblab/CircleNet/blob/4e0a8063d3591787e022d18427d258c3c20c237f/src/lib/trains/circledet.py#L17). 

> I believe that I should've initialized the weights from pre-trained model by having `load_model path/to/ctdet_coco_hg.pth` , where ctdet_coco_hg.pth was downloaded from [this page](https://github.com/hrlblab/CircleNet/blob/master/docs/GETTING_STARTED.md). Please let me know if I'm missing something. Thank you!

Yes, that's right.
